src/node_perf.cc

```diff
@@ -548,71 +562,65 @@ bool ELDHistogram::Disable() {
   return true;
 }
 
-// Event Loop Timing Histogram
-namespace {
-void RWLHistogramMin(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
-  Histogram::GetMin(env->req_wrap_latency(), args);
-}
-
-void RWLHistogramMax(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
-  Histogram::GetMax(env->req_wrap_latency(), args);
-}
-
-void RWLHistogramMean(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
-  Histogram::GetMean(env->req_wrap_latency(), args);
-}
-
-void RWLHistogramExceeds(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
-  Histogram::GetExceeds(env->req_wrap_latency(), args);
-}
-
-void RWLHistogramCount(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
-  Histogram::GetCount(env->req_wrap_latency(), args);
-}
-
-void RWLHistogramStddev(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
-  Histogram::GetStddev(env->req_wrap_latency(), args);
-}
 
-void RWLHistogramPercentile(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
-  Histogram::GetPercentile(env->req_wrap_latency(), args);
-}
-
-void RWLHistogramPercentiles(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
-  Histogram::GetPercentiles(env->req_wrap_latency(), args);
+RWLHistogram::RWLHistogram(
+    Environment* env,
+    Local<Object> wrap) : HistogramBase(env, wrap, 1, 3.6e12),
+    providers_(0) {}
+
+bool RWLHistogram::Enable() {
+  if (!enabled_) {
+    enabled_ = true;
+    env()->add_req_wrap_listener(this);
+    return true;
+  }
+  return false;
 }
 
-void RWLHistogramEnable(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
-  bool enabled = env->req_wrap_latency_enabled();
-  if (!enabled) {
-    env->req_wrap_latency()->Reset();
-    env->set_req_wrap_latency_enabled(true);
+bool RWLHistogram::Disable() {
+  if (enabled_) {
+    enabled_ = false;
+    env()->remove_req_wrap_listener(this);
+    return true;
   }
-  args.GetReturnValue().Set(!enabled);
+  return false;
 }
 
-void RWLHistogramDisable(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
-  bool enabled = env->req_wrap_latency_enabled();
-  if (enabled)
-    env->set_req_wrap_latency_enabled(false);
-  args.GetReturnValue().Set(enabled);
+void RWLHistogram::Record(int type, int64_t delta) {
+  if (!providers_[type])
+    return;
+  Histogram::Record(delta);
+  if (*TRACE_EVENT_API_GET_CATEGORY_GROUP_ENABLED(
+          TRACING_CATEGORY_NODE2(perf, reqwrap)) != 0) {
+    auto value = tracing::TracedValue::Create();
+    DumpToTracedValue(value.get());
```

Comment:
I feel that it is an overkill to trace the whole Histogram on every call to `Record`.

Reply:
We can tweak this later if necessary. How often would be acceptable?

Reply:
Usually I would expect that traces act on raw data therefore trace `type` and `delta` on every call of `Record`.
I would not expect a trace holding the whole histogram in the same category as it is redundant information.

But as you said on summit the requirements what should be traced are missing in general so feel free to keep this as it is and adapt once there is a more clear picture in this.

Reply:
We could tweak this a bit to be sensitive to detail level... e.g. output type and delta for category `node.perf.reqwrap` and the full histogram on `node.perf.reqwrap.detail` or something similar, but for now, I'd like to leave it as is.